### PR TITLE
New module for azure virtual network gateway

### PR DESCRIPTION
**Pull request by Madhura-CSI**
##### SUMMARY
New module for azure virtual network gateway

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
azure_rm_virtualnetworkgateway

##### ANSIBLE VERSION

```
ansible 2.6.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/centos-dev/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### ADDITIONAL INFORMATION
Currently we don't have any existing module to create/update/delete virtual network gateway(VPN gateway) in azure cloud. Hence, I would like to contribute this module to community.
Sample module call is as follows:

```
    - name: Create virtual network gateway without bgp settings
      azure_rm_virtualnetworkgateway:
        resource_group: testrg
        name: testvpngw
        ip_configurations:
          - name: testipconfig
            private_ip_allocation_method: Dynamic
            public_ip_address_name: testipaddr
        virtual_network: testvnet
        tags:
          common: "xyz"
      register: vgw_without_bgp
```
By default module waits for VPN gateway to be created.(using poller)

### CODE REVIEW

**Comment by Madhura-CSI** [comment]
WIP

**Comment by Fred-sun** [comment]
@Madhura-CSI Thanks for the contribution, Could you help confirm the PR ready for review or not？ if yes, I will push to review! Thanks!

**Comment by Fred-sun** [comment]
@Madhura-CSI Thanks for the contribution. Could you help finished this PR change? when you finished, I will push to review. Thanks!

**Comment by Fred-sun** [comment]
@Madhura-CSI Thanks for your contribution, Are you still here？Could you help finish the PR changed? if you finished, I will push to review? Thanks!

**Comment by Madhura-CSI** [comment]
Yes @Fred-sun , I am working on this PR. Will need few more days to finish the stuff. I will update here once this is ready for review.

**Comment by Fred-sun** [comment]
@Madhura-CSI Thanks for your update， When you finished changed, I will test,. Thanks!

**Comment by Fred-sun** [comment]
kindly ping

**Comment by Fred-sun** [comment]
kindly ping

**Comment by zikalino** [comment]
@Madhura-CSI I would like to add a few things and also integration tests and merge this pr.
would it be ok for you to add me to your fork?
I would also do the same with local network gateway.

**Comment by Madhura-CSI** [comment]
Sure @zikalino . Please go ahead and make the necessary changes as per your requirement.

**Comment by zikalino** [comment]
@Madhura-CSI great! :-) could you add me as a collaborator to your fork?

**Comment by Madhura-CSI** [comment]
@zikalino  One problem I see here is: I think we need to add write access at repository/org level and not at the pull-request level. Can't really do that as of now. Do you see any probable alternative to this? I would be glad to assist.

**Comment by zikalino** [comment]
@Madhura-CSI ok then. no problem. I will just fork your fork and create PR on the top of your branch.
a bit more complex, but should work.

**Comment by Madhura-CSI** [comment]
@zikalino sure, thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        if self.check_mode:
+            return self.results
+        if changed:
+            if self.state == 'present':
+                if not self.sku:
+                    self.fail('Parameter error: sku is required when creating a vpn gateway')
+                if not self.ip_configurations:
+                    self.fail('Parameter error: ip_configurations required when creating a vpn gateway')
+                subnet = self.network_models.SubResource(
+                    '/subscriptions/{0}/resourceGroups/{1}/providers/Microsoft.Network/virtualNetworks/{2}/subnets/GatewaySubnet'.format(
```
Comment (zikalino) [inline lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py]: @Madhura-CSI i have just started working on this pr, and I have a quick question.
why is GatewaySubned hardcoded here. shouldnt' we refer any subnet?
any special reason?

**Comment by Fred-sun** [comment]
@Madhura-CSI Are you still here? Could you give a feedback for zikalino's comments? Thanks!

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py (hunk 1 above)
Comment (Madhura-CSI) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py]: @zikalino,  'GatewaySubnet' < this name is mandatory for the subnet where we deploy the virtual network gateway virtual machines. AFAIK, we cannot change this name. If we ever do that, network gateway creation will fail at the end. Please confirm this behavior.

**Comment by Madhura-CSI** [comment]
@Fred-sun replied to the code review comment above. Please check.

File: lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py (hunk 1 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py]: @Madhura-CSI thanks! I will create PR to your fork soon :-)

**Comment by Fred-sun** [comment]
@Madhura-CSI Thanks for your update.
@zikalino Once you create PR , please paste link at here! Thanks!

**Comment by zikalino** [comment]
@Madhura-CSI I have created a pr to your fork some time ago. could you take a look and merge it?

**Comment by Madhura-CSI** [comment]
sure @zikalino . I will update this in a day or two and will let you know.

**Comment by zikalino** [comment]
@Madhura-CSI could you try to merge changes today? the freeze is coming :-)
otherwise I may try to merge in a separate pr

**Comment by Madhura-CSI** [comment]
@zikalino , sorry for the delay. I have now merged your branch with this one. Please review and proceed with upstream merge if everything looks good to you.
Thanks!

**Review by nitzmahone — APPROVED** [review]

**Comment by MyronFanQiu** [comment]
@Madhura-CSI Hello. Thanks for this great PR. Recently, we met a small [issue](https://github.com/Azure/azure_preview_modules/issues/260), I just wonder, could you share your design purpose for the `subnet` option?
https://github.com/ansible/ansible/blob/a3c3fd91f8a6c46d85b690f832e130ec37354d9c/lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py#L75-L78
Based on the current implementation, it seems that this option is not used in the code now. I checked the sdk, the param of the same name is constructed by the information in the virtual network.
https://github.com/ansible/ansible/blob/a3c3fd91f8a6c46d85b690f832e130ec37354d9c/lib/ansible/modules/cloud/azure/azure_rm_virtualnetworkgateway.py#L313-L317
I think there should be some original reason for you to create this option. we can discuss about it and continue improving this module after our discussion.

**Pull request merged**